### DISCUSSION TITLE
Allow LINQ expressions without `from` part

### DISCUSSION THREAD

**Discussion by ByteEater-pl** [dotnet/csharplang#5233] [Language Ideas]
I've sometimes seen in other people's code and even used myself the pattern `from _ in "_"` to reach for LINQ syntax in situations where there's no obvious source collection. Use cases:

1. Expression-local variable binding: `let x = e1 select e2`
2. Shorter than `Enumerable.Single(x)` way to create an IEnumerable singleton: `select x`

Some lookahead may be needed. I haven't checked, but I suppose `from` is treated specially to disambiguate cases when it's just a variable name. The same would become necessary for `let`, `select` and a few others (and in VB, to which this feature request applies equally well, possibly even more, e.g. `Aggregate`, given its generally richer LINQ syntax).

The (1) is has been discussed in recent years in many other places on  dotnet/csharplang, particularly those related to destructuring and pattern matching.

**Comment by CyrusNajmabadi** [comment]
What does `e1 select e2` mean? What does it get translated to?

**Reply by ByteEater-pl** [reply]
It gets `from _ in "_" ` prepended (assuming `_` isn't a free variable in `e1` or `e2`; surely you get the idea). So:
```
from _ in "_" let x = e1 select e2

**Reply by CyrusNajmabadi** [reply]
but what are `e1` and `e2`?  I genuinely don't get teh purpose of this.  It doesn't look like `_` or `"..."` or 'x' are used at all... so what does thsi get you?

**Reply by ByteEater-pl** [reply]
`e1` and `e2` ere arbitrary expressions. `x` may be used in `e2`. The `_` not being used is the whole point. This issue is about eliminating the need for `from _ in "_"` as the artificial prefix to enable LINQ syntax without a source collection.

Many languages have similar constructs. In F# the equivalent would be:
```
let x = e1
[e2]
```
and in Haskell:
```
let x = e1 in pure e2
```

(Sorry for so many edits. The `select`s translating to `Select` or `SelectMany` got me confused after a few years of not using LINQ often.)

**Reply by CyrusNajmabadi** [reply]
i don't know what either of those mean.  Can you explain what semantics they have?  what purpose does it serve to do:

```c#
let x = e1
e2
```

here, 'x' isn't used.  so i don't see why even have that line.

```c#
let x = e1 in e2
```

What is 'x' here?  I don't understand what this is doing or what it translates to.

**Reply by ByteEater-pl** [reply]
`x` is an identifier. The semantics (static and dynamic in one go for brevity) is:
1. Evaluate `e1`.
2. Create a variable named `x` with the value obtained in (1) and scope enclosing `e2`.
3. Evaluate `e2`.
4. Call `Enumerable.Single` on the result of (3), giving the result for the whole expression.

That being LINQ, in more complex expressions, e.g. with `from` after `let`, some extension methods (like `Select`, `SelectMany`, `Where`, `GroupBy`) may get called, as specified in the LINQ semantics.

**Reply by CyrusNajmabadi** [reply]
I guess i don't see why this needs to be done at all.  If 'x' isn't even used.  Why not just simply write this as:

```c#
e1;
```

Why is this put in a linq expression at all?

**Reply by ByteEater-pl** [reply]
That would be `e2`, not `e1`. But again, `e1` and `e2` are placeholders for arbitrary expressions. There may indeed be free occurrences of `x` in `e2`.

**Reply by CyrusNajmabadi** [reply]
> That would be e2, not e1

No, i meant e1'.  You have: 

```c#
let x = e1
```

if 'x' wasn't used, why not just write `e1;`?

> There may indeed be free occurrences of x in e2.

can you provide an example that then does that :)   i'm basing my judgement here based on the examples given.  

in the case where 'x' is used, then i don't see why you need `let`.  Why not just write:

```c#
var x = e1;
/* expr using x */
```
?

**Reply by ByteEater-pl** [reply]
> No, i meant e1'.

OK, then I misunderstood, sorry.

> Why not just write:
>
```
var x = e1;
/* expr using x */
```
> ?

It's a series of statements. These don't fit in an expression context. (There obvious alternative is IIFE (immediately invoked function expression) – a lambda expression as the callee in a call expression (with appropriate casting) or with Invoke called on it. But that's cumbersome in C#.)

> can you provide an example that then does that

I actually happened to see one yesterday in some discussion on dotnet/csharplang, can't find it now however.
Another example then:
```
let q = -b / a / 2 let d = Math.Sqrt(q * q - c / a) select new [] {q + d, q - d} // roots of quadratic polynomial

**Reply by CyrusNajmabadi** [reply]
> It's a series of statements. 

I don't get why that's an issue.

> These don't fit in an expression context.

Why is that a bad thing?  

> `let q = -b / a / 2 let d = Math.Sqrt(q * q - c / a) select new [] {q + d, q - d} // roots of quadratic polynomial`

This genuinely seems.  strange.  It could jsut be written as:

```c#
var q = -b / a / 2;
var d = Math.Sqrt(q * q - c / a);
var roots = (q + d, q - d);
```

I genuinely see no reason to use linq for this...

**Reply by DavidArno** [reply]
@CyrusNajmabadi,

This technique can be used to work around the fact that switch expressions cannot support statements. So I can do something like this instead:
```cs
var result = (1, 2.0, 3.0, 4.0) switch {
    (1, var a, var b, var c) => from _ in "_" 
                                let q = -b / a / 2 
                                let d = Math.Sqrt(q * q - c / a) 
                                select new [] {q + d, q - d},
    _ => null
};
```

Removing the need for `from _ in "_" ` makes it slightly tidier. But it's still a hack in my view. I feel that a far neater approach is to push that code into a function, so we get something like:
```cs
var roots = (double a, double b, double c) => {
    var q = -b / a / 2;
    var d = Math.Sqrt(q * q - c / a);
    return (q + d, q - d);
};

var result = (1, 2.0, 3.0, 4.0) switch {
    (1, var a, var b, var c) => roots(a, b, c),
    _ => (-1, -1)
};
```
And for those that don't like this approach, the language team are quite likely to implement support for statements in switch expressions. So in future it could be written as something like:
```cs
var result = (1, 2.0, 3.0, 4.0) switch {
    (1, var a, var b, var c) => {
        var q = -b / a / 2;
        var d = Math.Sqrt(q * q - c / a);
        (q + d, q - d) },
    _ => (-1, -1)
};
```
So I really don't see any benefit to allowing linq expressions without the from part.

**Comment by svick** [comment]
Is this effectively a different syntax for sequence expressions (https://github.com/dotnet/csharplang/issues/377)? Using those, you would write `(var x = e1; e2)` instead of `let x = e1 select e2`.

**Reply by ByteEater-pl** [reply]
More or less. The more being the availability of LINQ syntax. I remember doing `from _ in "_" let x = e1 from y in e2 select e3`.

**Reply by ByteEater-pl** [reply]
Another great example in this area of language design is XQuery whose [FLWOR expressions](https://www.w3.org/TR/2017/REC-xquery-31-20170321/#id-flwor-expressions) have the following grammar:
```
[41] | FLWORExpr | ::= | InitialClause                                 IntermediateClause*  ReturnClause |  
-- | -- | -- | -- | --
[42] | InitialClause | ::= | ForClause  \|  LetClause  \|  WindowClause

**Reply by CyrusNajmabadi** [reply]
> The more being the availability of LINQ syntax.

As above, i don't understand why that's desirable (or even the expected semantics).
